### ISSUE TITLE
Set camera mask for new elements

### ISSUE THREAD

**Issue by jonyrock** [cocos2d/cocos2d-x#13203]
Hi

It is not obvious that I should write every time such thing

``` C++
auto sprite = //<some new 3d spite>
sprite->setCameraMask(getCameraMask());
addChild(sprite);
```

I think this behaviour should be by default

**Comment by zilongshanren** [comment]
@super626  
Could you take a look at this issue?

**Comment by stevetranby** [comment]
We just had this discussion come up again. We think that the sprite should get added with the parent node's camera mask. Maybe have a flag as whether to cascade or not. I would guess there's only a few, or none, instances where someone would want to addChild(node) while setting a different camera mask than it's parent.

I can submit a PR, but wanted to see if there's any discussion first.

```
// in method => void Node::addChildHelper(...) { 
child->setParent(this);
child->setCameraMask(getCameraMask()); // ADDED
```

http://discuss.cocos2d-x.org/t/solved-camera-not-applied-to-newly-created-object-sprite/29323/

**Comment by minggo** [comment]
I think we need it. @ricardoquesada what's your opinion?

**Comment by ricardoquesada** [comment]
yes, sounds good.

**Comment by minggo** [comment]
@stevetranby will you please send a PR?

**Comment by stevetranby** [comment]
Sure, been out on vacation. Will submit by Monday.

**Comment by minggo** [comment]
thanks @stevetranby

**Comment by stevetranby** [comment]
https://github.com/cocos2d/cocos2d-x/pull/16094

**Comment by ricardoquesada** [comment]
@stevetranby this patch broke the Scene3dTest. No nodes are shown on that test (cpp-tests -> Scene3dTest)

Any idea how improve this patch so that it works with the Scene3dtest ?

Perhaps this patch if fixing a real bug, but I'm worried that perhaps some games are relying on this bug... like Scene3dtest.

@minggo any idea what should we do?

Thanks.

**Comment by stevetranby** [comment]
@ricardoquesada ah, interesting, I wonder if just testing if the child's mask is currently set? I presume it's camera default at construction, so maybe if default then set as parent? I'm sure that's still wrong, with respect to anything similar to Scene3DTest where default camera may be used in the same way as the other cameras.

Maybe it's the wrong idea for a default as well, though we've found ourselves setting it explicitly in many places until we figured out we really just want addChild and setParent to set the same mask.

I'll take a look at that test and see if it sheds any light on this.

Prob smart to just revert for now?

**Comment by stevetranby** [comment]
Yeah, being reminded that the set mask cascades it should probably at minimum send in false for the optional cascade param. I'll continue to look through that test to understand why child is setting different camera mask than its parent.

**Comment by ricardoquesada** [comment]
@stevetranby ok. thanks.
I'll keep the PR applied for now. If we can't find a better solution, I'll revert it. thanks.

**Comment by stevetranby** [comment]
(sorry a bit long for such a minor change, I could move/copy over to forums if desired)

@ricardoquesada @minggo 

After a little investigation I've noticed a few use cases with different ideas of how to use cameras and their masks. This is a common case of myself understanding some 'pattern' and thinking it makes the most logical sense and not thinking of others.

HOWEVER, I would also admit that it is unexpected to have a previously set camera mask change to the one that the parent node has when adding as a child. So in the strict sense reverting this PR is probably more correct, but I would say requires more work if using multiple cameras effectively.

**If this were my engine I would probably keep this PR, change to not cascade in addChild, and also consider having it not set the mask to the parent's if the child's mask is not the default.**

_Could also have a -1 UNKNOWN value. UNKNOWN and DEFAULT would render on DEFAULT camera, but would allow testing whether camera mask has been set or not._

### Pattern [PR reverted]

Set the camera mask on all nodes that need a mask OR add all children and then set the mask on various "root" nodes using the default cascading mask onto its children. For runtime additions set the mask manually for each or (again) only on some common parent node with cascade. 

### Pattern [PR applied]

Set the camera mask on various "root" nodes. Add all children without any setting any camera mask. Then set any nodes with custom (different) camera mask after they've been added to the scene (addChild).

Our game, for example, uses this process now and then when we display a dialog it has a root node (customDialogNode) that can have its camera mask set to HUD. 

**I'm still not sure the correct default for everyone. As a biased viewpoint I think that WITH PR requires less code written by everyone in the long-run following the recommended way.**

---

### NavMeshTest.cpp

This test shows an example shows how setCameraMask is often used on every child added to the scene.

**With this PR applied the camera mask would need to only be set on the "root" parent node that contains all the children with matching masks.**

### Scene3DTest.cpp

This test shows an example where nodes with different camera masks are added as children to the same node (parent) which has no set camera mask (default). In this case it is wrong for `addChild` to change the camera mask. 

**With the PR applied one could apply custom/different camera masks to encapsulated nodes or set the custom camera masks after adding each child.**

```
// with PR applied (no need to set component's mask if same as parent)
// set "root" parent to the most common camera mask
playerBg->setCameraMask(s_CM[LAYER_TOP]);
playerBg->addChild(....); // for each node
// set custom masks after addChild(....) for that node
girl->setCameraMask(s_CM[LAYER_MIDDLE]);
// no longer necessary
//zoomIn->setCameraMask(s_CM[LAYER_TOP]);
//zoomOut->setCameraMask(s_CM[LAYER_TOP]);
//slider->setCameraMask(s_CM[LAYER_TOP]);
```

**Comment by minggo** [comment]
@stevetranby @ricardoquesada if we can not define which behavior is a correct, then keep original implementation is better.

**Comment by minggo** [comment]
@ricardoquesada @stevetranby i think we should revert it since we can not define `correct` behavior and it breaks compatibility.

**Comment by ricardoquesada** [comment]
@stevetranby Thanks. Makes what you said. Do you think there is a way to make it work the way you want and also making it backwards compatible?

I noticed that the method `Node::setCameraMask()` supports an additional argument `applyChildren`, which will recursively will apply the mask to all its current children.

something like this:

``` c++
girl->setCameraMask(s_CM[LAYER_TOP], true);
```

Not sure if such argument will help in this scenario, but apparently we are not using it in our tests.

**Comment by stevetranby** [comment]
@ricardoquesada so the second argument is true by default and we never specify false mostly because either it's being applied to some root parent node and we want it to cascade or the node has no children so it doesn't matter. 

This PR is just a convenience default and my only idea that I think might work that would be "backward compatible" is to have an UNSET mask at construction and only apply the parent's mask if the child's mask is UNSET (rendering would have to treat DEFAULT and UNSET the same, as I mentioned previously).

Let's just revert it. And it should also be possible for me to create a new PR to remove any redundant calls to setCameraMask where possible.

In the end too few people are probably even using cameras, much less care about a few extra lines of code.

**Comment by minggo** [comment]
I reverted it. So let's open the issue again to find a way that may not break compatibility.

**Comment by cezheng** [comment]
Yep, actually we also reverted this commit on our fork which being used by our project.
https://github.com/cezheng/cocos2d-x/commit/b1e77a209233854cb8dcdb59e9434ff74b1ac4c1

maybe adding a addChild method with a extra `useParentCameraFlag` should do the job.

**Comment by stevetranby** [comment]
@cezheng that'd prob be the one backward compatible way to add it. I'm not a huge fan of further overloading addChild, however, especially when most use cases seem to stick to passing in the child node and leaving the rest as default values. It's most likely easier to set the mask separately even if there were a fourth (4th) parameter option for the mask. 

The best path forward until v4 (or never) is probably to just improve the couple tests so that they use the setCameraMask method as little as possible by taking into account that it cascades (applies to all children). 

Maybe mention in the forum post that one should derive Node/Sprite or create a global method if you want to auto-set the mask to the parent.

```
GameObject::addChildWithMask(child, mask);
utils::addChildWithMask(parent, child, mask);
```

The normal naive use case is to always set the mask. If one sets the mask only on key parent nodes or inside a factory create function then there's no real need for this Issue's idea.

And thinking about the idea of using a default NONE/UNKNOWN mask is not ideal either since one might want to compare and expect DEFAULT most of the time.

If this issue/idea(s) is taken into account when designing the API and architecture I think it could be improved slightly, though maybe not since for example Unity handles this through the editor (mostly) where you choose camera masks using the GUI editor dropdown selector.
